packages/editor/src/components/post-publish-panel/maybe-upload-media.js

```diff
@@ -0,0 +1,169 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	PanelBody,
+	Button,
+	Spinner,
+	__unstableMotion as motion,
+	__unstableAnimatePresence as AnimatePresence,
+} from '@wordpress/components';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+import { upload } from '@wordpress/icons';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+import { useState } from '@wordpress/element';
+import { isBlobURL } from '@wordpress/blob';
+
+/**
+ * Internal dependencies
+ */
+import { store as editorStore } from '../../store';
+
+function flattenBlocks( blocks ) {
+	const result = [];
+
+	blocks.forEach( ( block ) => {
+		result.push( block );
+		result.push( ...flattenBlocks( block.innerBlocks ) );
+	} );
+
+	return result;
+}
+
+function Image( block ) {
+	const { selectBlock } = useDispatch( blockEditorStore );
+	return (
+		<motion.img
+			tabIndex={ 0 }
+			role="button"
+			aria-label={ __( 'Select image block.' ) }
+			onClick={ () => {
+				selectBlock( block.clientId );
+			} }
+			onKeyDown={ ( event ) => {
+				if ( event.key === 'Enter' || event.key === ' ' ) {
+					selectBlock( block.clientId );
+					event.preventDefault();
+				}
+			} }
+			key={ block.clientId }
+			alt={ block.attributes.alt }
+			src={ block.attributes.url }
+			animate={ { opacity: 1 } }
+			exit={ { opacity: 0, scale: 0 } }
+			style={ {
+				width: '36px',
+				height: '36px',
+				objectFit: 'cover',
+				borderRadius: '2px',
+				cursor: 'pointer',
+			} }
+			whileHover={ { scale: 1.08 } }
+		/>
+	);
+}
+
+export default function PostFormatPanel() {
+	const [ isUploading, setIsUploading ] = useState( false );
+	const externalImages = useSelect( ( select ) => {
+		const { getEditorBlocks } = select( editorStore );
+		return flattenBlocks( getEditorBlocks() ).filter( ( block ) => {
```

Comment:
I think it's better to flatten and filter outside the useSelect to avoid return new array instances, no?

Reply:
Nice catch!